File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsViewState.java
package com.example.android.architecture.blueprints.todoapp.statistics;

import android.support.annotation.Nullable;
import com.example.android.architecture.blueprints.todoapp.mvibase.MviViewState;
import com.google.auto.value.AutoValue;

@AutoValue abstract class StatisticsViewState implements MviViewState {
  abstract boolean isLoading();

  abstract int activeCount();

  abstract int completedCount();

  @Nullable abstract Throwable error();

  public abstract Builder buildWith();

  static StatisticsViewState idle() {
    return new AutoValue_StatisticsViewState.Builder().isLoading(false)
        .activeCount(0)
        .completedCount(0)
        .error(null)
        .build();
  }

  @AutoValue.Builder static abstract class Builder {
    abstract Builder isLoading(boolean isLoading);

    abstract Builder activeCount(int activeCount);

    abstract Builder completedCount(int completedCount);

    abstract Builder error(@Nullable Throwable error);

    abstract StatisticsViewState build();
  }
}

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/StatisticsViewState.kt
package com.example.android.architecture.blueprints.todoapp.statistics

import com.example.android.architecture.blueprints.todoapp.mvibase.MviViewState

data class StatisticsViewState(
    val isLoading: Boolean,
    val activeCount: Int,
    val completedCount: Int,
    val error: Throwable?
) : MviViewState {
  companion object {
    fun idle(): StatisticsViewState {
      return StatisticsViewState(
          isLoading = false,
          activeCount = 0,
          completedCount = 0,
          error = null
      )
    }
  }
}
